svgnative/ports/cairo/CairoImageInfo.c

```diff
@@ -0,0 +1,149 @@
+/*
+Copyright 2019 suzuki toshiya <mpsuzuki@hiroshima-u.ac.jp>. All rights reserved.
+This file is licensed to you under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License. You may obtain a copy
+of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software distributed under
+the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+OF ANY KIND, either express or implied. See the License for the specific language
+governing permissions and limitations under the License.
+*/
+
+#include "CairoImageInfo.h"
+#include <string.h>
+
+/* JPEG (image/jpeg)
+ */
+#include <stdio.h>
+#include <stdlib.h>
+#include <setjmp.h>
+#include <jpeglib.h>
+
+struct _cairo_jpeg_error_mgr
+{
+    struct jpeg_error_mgr pub;
+    jmp_buf setjmp_buffer;
+};
+
+typedef struct _cairo_jpeg_error_mgr * _cairo_jpeg_error_ptr;
+
+METHODDEF(void)
+_cairo_jpeg_error_exit (j_common_ptr cinfo)
+{
+    _cairo_jpeg_error_ptr _cairo_jpeg_err = (_cairo_jpeg_error_ptr) cinfo->err;
+    (*cinfo->err->output_message) (cinfo);
+    longjmp(_cairo_jpeg_err->setjmp_buffer, 1);
+}
+
+cairo_surface_t *
+_cairo_image_surface_create_from_jpeg_stream(const unsigned char* data,
+                                             unsigned int length)
+{
+    cairo_surface_t* _cairo_jpeg_surface = NULL;
+
+    struct jpeg_decompress_struct cinfo;
+    struct _cairo_jpeg_error_mgr jerr;
+    cinfo.err = jpeg_std_error(&jerr.pub);
+    jerr.pub.error_exit = _cairo_jpeg_error_exit;
+
+    unsigned char* outBuff = NULL;
+    if (setjmp(jerr.setjmp_buffer)) {
+        jpeg_destroy_decompress(&cinfo);
+        if (_cairo_jpeg_surface) {
+            cairo_surface_destroy (_cairo_jpeg_surface);
+        };
+        return NULL;
+    };
+
+    jpeg_create_decompress(&cinfo);
+    jpeg_mem_src(&cinfo, data, length);
+    jpeg_read_header(&cinfo, TRUE);
+
+    cairo_format_t cairo_color_format;
+    switch (cinfo.out_color_space) {
+    case JCS_GRAYSCALE:
+        cairo_color_format = CAIRO_FORMAT_A8;
+        break;
+    case JCS_EXT_ARGB:
+        cairo_color_format = CAIRO_FORMAT_ARGB32;
+        break;
+    case JCS_RGB:
+    default:
+        cairo_color_format = CAIRO_FORMAT_RGB24;
+    };
+
+    /* cinfo.image_width, cinfo.image_height, cinfo.num_components are already filled, but
+     * cinfo.output_width, cinfo.output_height, cinfo.output_components are not, because
+     * they are output parameters
+     */
+    jpeg_start_decompress(&cinfo);
+
+    size_t outCur, outLimit;
```

Comment:
default initialize variables.

Reply:
I would move them to their first usages. Also I would add a comment saying as this C source requires C99 compilers. In future, maybe I would try to update this source for C89-compatible...